community,wg,code-quality: establish working group code-quality (#333)
kubevirt/community: wg-code-quality/wg-charter.md

# WG Code-Quality Charter

## Goals

The Code-Quality Working Group has two goals:

* (short-term) **focus on the improvement of code quality** and
* (long-term) **create an inter-SIG process for identifying, agreeing on and addressing code quality issues**

The long-term goal is achieved when the process is implemented and executed as required.

## Definition of objectives

The main objective of this effort is to improve the quality of the project codebase.

With this in mind, there is a need to define a clear scope on which areas and through what means this can be accomplished.

* Create and maintain code-quality guidelines, best practices and standards.
* Assure distribution of knowledge in regard to code quality.
  * This may include presentations, documentation, lectures, certifications, linters, etc.
* Enforcement of selected standards.
  * Through automation and active intervention in code reviews.

>   * Through codebase infrastructure.

Reviewer: ```suggestion
  * Through codebase infrastructure. E.g reusable libraries.
```
(the existing text was not clear to me without the e.g)

* Management:
  * Prioritizing tasks.
  * Focusing on selected topics.

> * Monitor and identification of code-quality issues.

Reviewer: ```suggestion
* Monitoring and identification of code-quality issues.
```

## Deliverables

> The inter-SIG process definition will be manifested in a document that defines the process itself, also how and when it is executed. 

Reviewer 2: I think the deliverables should be the result of the scope points (objectives in the current naming), more or less.

I like the general wording of the device-manager WG:

> The WG will coordinate the delivery of KEPs and their implementations by the participating SIGs. Interim artifacts will include documents capturing use cases, requirements, and designs; however, all of those will eventually result in KEPs and code owned by SIGs.

We could re-word it a bit to fit better to our naming (or lack of it) and scope:

> The WG will coordinate the prioritization and delivery of design-proposals, documentation, implementations and other content that aims to improve the code quality of the project.

# WG Code-Quality Charter

## Scope

The Code-Quality Working Group has two goals:

* (short-term) **focus on the improvement of code quality** and

> * (long-term) **create an inter-SIG process for identifying, agreeing on and addressing code quality issues**

Reviewer 2: For me, line 14 is summarizing this and describe why the WG is needed.
The long-term goal here, which mentions a processes, is not very clear to me.
I would expect the WG to have artifacts from which the quality will be improved and sustained over time.

This may be partially achieved through processes, but I am unsure if that is the "solution" or how it will look-like.

Author: I have changed the In Scope section by changing the last sentence to what you pointed out before, which IMHO brings it down into an actionable goal. WDYT?

Reviewer 2: I guess I do not see this inter-SIG process in my mind and therefore have hard time to understand and agree on it being a goal. I feel this section is redundant as everything is described in the next section.

However, this is a small part compared to all this charter content, so I should not get stuck on it.
We can raise it later and adjust per need.

The long-term goal is achieved when the process is implemented and executed as required.

## In scope

The main objective of this effort is to continuously improve the quality of the project codebase.

With this in mind, there is a need to define a clear scope on which areas and through what means this can be accomplished.

* Create and maintain code-quality guidelines, best practices and standards.
* Assure distribution of knowledge in regard to code quality.
  * This may include presentations, documentation, lectures, certifications, linters, etc.
* Enforcement of selected standards.
  * Through automation and active intervention in code reviews.
  * Through codebase infrastructure. E.g reusable libraries.
* Management:
  * Prioritizing tasks.
  * Focusing on selected topics.
* Monitoring and identification of code-quality issues.

> The WG will therefore work on processes through which the above goals can be achieved.

Reviewer 2: > work on processes

I am unsure if this is clear enough to keep it in the text.

As I see it, the WG should work on providing the means to improve the quality until a point where it is well integrated in the projects. The points above hints what we expect to have in the end.

I do not think it is limited to creating processes, it is there to facilitate the effort to create content, refactor code, etc. As it is already mentioned in the points above, I think this sentence is redundant.

Author: I have changed the In Scope section by changing the last sentence to what you pointed out before, which IMHO brings it down into an actionable goal. WDYT?

# WG Code-Quality Charter

## Scope

The Code-Quality Working Group has two goals:

> * (short-term) **focus on the improvement of code quality** and

Reviewer 3: nit - drop the `and` at the end

* (long-term) **create an inter-SIG process for identifying, agreeing on and addressing code quality issues**

The long-term goal is achieved when the process is implemented and executed as required.

## In scope

> The main objective of this effort is to continuously improve the quality of the project codebase.

Reviewer 4: Since the project is quite big, is there an idea of where to start?

Reviewer 2: The charter is focused on *what* is needed rather on *how*.
The definition of the scope just provides the understanding of what the WG is considering under its influence/scope.

As @dhiller pointed out, we will discuss how to do things and which areas to invest first in our bi-weekly meetings and through regular communications.

With this in mind, there is a need to define a clear scope on which areas and through what means this can be accomplished.

> * Create and maintain code-quality guidelines, best practices and standards.

Reviewer 3: nit - link to https://github.com/kubevirt/kubevirt/blob/main/docs/coding-conventions.md ?

Reviewer 2: The link points to *how" which I think should be left out of a charter.
Here we just want to say that creating and maintaining such a convention is in the WG scope, that is it.

* Assure distribution of knowledge in regard to code quality.
  * This may include presentations, documentation, lectures, certifications, linters, etc.
* Enforcement of selected standards.
  * Through automation and active intervention in code reviews.
  * Through codebase infrastructure. E.g reusable libraries.
* Management:
  * Prioritizing tasks.
  * Focusing on selected topics.
* Monitoring and identification of code-quality issues.

The WG will work on facilitating and managing the quality effort until enough tooling, documentation, code structure, reviewer and maintainer awareness is achieved.

## Stakeholders

### Special Interest Groups

* All SIGS claiming code ownership and producing code, currently

>   * SIG Compute,
>   * SIG Network and
>   * SIG Storage
> * sig-ci

Reviewer 4: nit:
```suggestion
  * SIG Compute
  * SIG Network
  * SIG Storage
  * SIG-CI
```

  * automation purposes of code-quality measures

### Roles and responsibilities

#### Chairs [1]
- @EdDev
- @iholder101

#### Liaisons

> - @dhiller

Reviewer 3: This isn't important but it might be useful to have someone from `sig-storage` as a chair or liaison?

Reviewer 2: It will be great to have another chair, 3 sounds ideal.
The WG needs to communicate with other SIGs to apply some of its work items, however, we do not require representatives from SIGs to have a role in the WG. At least I do not see a reason for that.

Author: (also since I am quite deeply involved in community matters, I am willing to help out to that regard)

Reviewer 3: This was just a suggestion to help with initial communication out of the WG to the final active SIG in the list. As @EdDev said this isn't required and once things start moving forward with the WG I'm sure formal communication to all of the SIGs will be more than adequate.

## Deliverables

The inter-SIG process definition will be manifested in one or more documents that define the quality effort:

> * what tooling and what documentation is required,
> * what metrics for code structure should be established and how it will be implemented,

Reviewer 4: nit:
```suggestion
* what tooling and what documentation is required
* what metrics for code structure should be established and how it will be implemented
```

* how reviewer and maintainer awareness will be achieved

Also those documents will include how and when effective measures are executed.

## Timelines and Disbanding

> The WG will decide at a certain point that the basic goal is achieved - when there is good enough documentation, tooling and metrics about the targets described above available. In that case it can decide to disband itself.

Reviewer 5: I think that this kind of phrasing could be a good compromise / starting point.
From my POV, let's achieve the goals above and then re-asses whether we still need this WG around or not. We might choose to disband it, or change its goal. As an example, lowering the meeting frequency from bi-weeky to monthly could make sense.

I think this shouldn't be a blocker for creating this WG.

Author: I agree that a fundamental part of this effort is the continuous improvement and refinement of the goals - this will also define the outcomes at large.

# WG Code-Quality Charter

## Scope

The Code-Quality Working Group has two goals:

* (short-term) **focus on the improvement of code quality**
* (long-term) **create an inter-SIG process for identifying, agreeing on and addressing code quality issues**

The long-term goal is achieved when the process is implemented and executed as required.

## In scope

The main objective of this effort is to continuously improve the quality of the project codebase.

With this in mind, there is a need to define a clear scope on which areas and through what means this can be accomplished.

* Create and maintain code-quality guidelines, best practices and standards.
* Assure distribution of knowledge in regard to code quality.
  * This may include presentations, documentation, lectures, certifications, linters, etc.

> * Enforcement of selected standards.
>   * Through automation and active intervention in code reviews.
>   * Through codebase infrastructure. E.g reusable libraries.

Reviewer 6: ```suggestion
* Enforcement of selected standards via the following methods:
  * Automation and active intervention in code reviews.
  * Codebase infrastructure. E.g reusable libraries.
```

* Management:
  * Prioritizing tasks.
  * Focusing on selected topics.
* Monitoring and identification of code-quality issues.

The WG will work on facilitating and managing the quality effort until enough tooling, documentation, code structure, reviewer and maintainer awareness is achieved.

## Stakeholders

### Special Interest Groups

> * All SIGS claiming code ownership and producing code, currently

Reviewer 6: ```suggestion
* All SIGs claiming code ownership and producing code, currently
```

  * SIG Compute
  * SIG Network

>   * SIG Storage

Reviewer 5: nit: observability? testing? documentation?
which other SIGs do we currently have? :D

Reviewer 6: This makes sense for now, based on SIGs with current charters, although SIG-observability has an [open PR](https://github.com/kubevirt/community/pull/353) so it would be worth including them.
Is the intention to update this as other SIGs are formalised?

> * sig-ci

Reviewer 6: ```suggestion
* SIG CI
```
